File: sdk/typescript/src/request.ts
import { AegisConfigurationError } from "./errors.js";
import { CapabilitiesRequest, coerceCapabilitiesPayload } from "./capabilities.js";
import { IntentContract, coerceIntentPayload } from "./intent.js";

export interface ExecutionRequestInit {
  language: string;
  code: string;
  timeoutMs?: number;
  intent?: IntentContract | Record<string, unknown>;
  capabilities?: CapabilitiesRequest | Record<string, unknown>;
  profile?: string;
  workspaceId?: string;
  executionId?: string;
}

export class ExecutionRequest {
  readonly language: string;
  readonly code: string;
  readonly timeoutMs?: number;
  readonly intent?: IntentContract | Record<string, unknown>;
  readonly capabilities?: CapabilitiesRequest | Record<string, unknown>;
  readonly profile?: string;
  readonly workspaceId?: string;
  readonly executionId?: string;

  constructor(init: ExecutionRequestInit) {
    if (init.intent !== undefined && init.capabilities !== undefined) {
      throw new AegisConfigurationError("provide either intent or capabilities, not both");
    }
    this.language = init.language;
    this.code = init.code;
    this.timeoutMs = init.timeoutMs;
    this.intent = init.intent;
    this.capabilities = init.capabilities;
    this.profile = init.profile;
    this.workspaceId = init.workspaceId;
    this.executionId = init.executionId;
  }

  toPayload(): Record<string, unknown> {
    const payload: Record<string, unknown> = { lang: this.language, code: this.code };
    if (this.timeoutMs !== undefined) payload.timeout_ms = this.timeoutMs;
    if (this.profile) payload.profile = this.profile;
    if (this.workspaceId) payload.workspace_id = this.workspaceId;
    if (this.executionId) payload.execution_id = this.executionId;
    const intentPayload = coerceIntentPayload(this.intent);
    const capabilitiesPayload = coerceCapabilitiesPayload(this.capabilities);
    if (intentPayload) payload.intent = intentPayload;
    if (capabilitiesPayload) payload.capabilities = capabilitiesPayload;
    return payload;
  }
}

File: sdk/python/aegis/request.py
from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any

from .capabilities import CapabilitiesRequest, coerce_capabilities_payload
from .errors import AegisConfigurationError
from .intent import IntentContract, coerce_intent_payload


@dataclass(slots=True)
class ExecutionRequest:
    language: str
    code: str
    timeout_ms: int | None = None
    intent: IntentContract | Mapping[str, Any] | None = None
    capabilities: CapabilitiesRequest | Mapping[str, Any] | None = None
    profile: str | None = None
    workspace_id: str | None = None
    execution_id: str | None = None

    def __post_init__(self) -> None:
        if self.intent is not None and self.capabilities is not None:
            raise AegisConfigurationError("provide either intent or capabilities, not both")

    def to_payload(self) -> dict[str, Any]:
        payload: dict[str, Any] = {"lang": self.language, "code": self.code}
        if self.timeout_ms is not None:
            payload["timeout_ms"] = self.timeout_ms
        if self.profile:
            payload["profile"] = self.profile
        if self.workspace_id:
            payload["workspace_id"] = self.workspace_id
        if self.execution_id:
            payload["execution_id"] = self.execution_id
        intent_payload = coerce_intent_payload(self.intent)
        capabilities_payload = coerce_capabilities_payload(self.capabilities)
        if intent_payload is not None:
            payload["intent"] = intent_payload
        if capabilities_payload is not None:
            payload["capabilities"] = capabilities_payload
        return payload
